src/librustc/middle/infer/mod.rs

```diff
@@ -613,6 +613,36 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         self.commit_unconditionally(move || self.try(move |_| f()))
     }
 
+    /// Execute `f` and commit the region bindings if successful
```

Comment:
We probably want a more detailed comment here. "...commit only the region bindings if successful. The function `f` must be very careful not to leak any non-region variables that get created."